src/core/public/workspace/workspaces_service.ts

```diff
@@ -26,8 +28,19 @@ export class WorkspacesService implements CoreService<WorkspacesSetup, Workspace
   private setFormatUrlWithWorkspaceId(formatFn: WorkspacesStart['formatUrlWithWorkspaceId']) {
     this.formatUrlWithWorkspaceId = formatFn;
   }
-  public async setup({ http }: { http: HttpSetup }) {
+  public async setup({ http, uiSettings }: { http: HttpSetup; uiSettings: IUiSettingsClient }) {
     this.client = new WorkspacesClient(http);
+
+    // If workspace was disabled while opening a workspace url, navigate to basePath
+    if (uiSettings.get('workspace:enabled') === false) {
```

Comment:
Do we need to disable the url rewrite in the server side?

Reply:
Right, ideally I would disable it, but I didn't find a good way to do that. Right now, when workspace is disabled, the UI won't load, and also when opening `/w/<id>` it will redirect user to home page. So I feel that might be enough. What do you think? cc @wanglam @SuZhou-Joe

Reply:
I'd vote for that we should disable the rewrite logic in server side when the flag is turned off.

Reply:
We can add a if else inside server.workspace_service.ts, the proxyToRealHandler method could consume the flag.

Reply:
@SuZhou-Joe What's your concern if we leave proxyToRealHandler as it is?

Reply:
Not concern, just thought it would be better to make OSD behaves exactly as what it does before Trinity if the flag is turned off.

Reply:
@SuZhou-Joe I updated the code so that it return 404 if access workspace path when workspace is disabled. This is the same behavior as now when visiting an unknown path.